thebeetoken/beenest-web
# Integrate token swaps at time of payment

## Description
This integrates Uniswap (via the [Unipay](https://github.com/woeltjen/unipay) "Payment Agent" contract) into the payment flow to allow BEE payments which originate in DAI.

Tracking to-dos:

- [x] Show DAI as an option
- [x] Get pricing at run-time
- [x] Call Uniswap during payment
- [x] Provide messaging during payment steps
- [x] Add margin for precision, price fluctuation
- [x] Figure out why guestConfirmBooking isn't being called...
- [x] Make sure guestConfirmBooking advances booking state correctly
- [x] Handle case when liquidity is just unavailable
- [x] Verify that `invoice` works
- [x] Fix test failures
- [x] Code cleanup? (Itemize future refactors?)

## How to Test
1. Get some "Ropsten DAI" (really just an ad hoc ERC-20 token)
    1. You can `mint` all you like on [etherscan](https://ropsten.etherscan.io/address/0x909e4dbdef114c9b39078df314177780b89d8062#writeContract)
2. Go make a booking
    1. There is limited Ropsten BEE liquidity; prefer cheap listings (e.g. actual nest of bees)
    2. (Or, find a very expensive booking and verify that payment is disabled with useful messaging...)
3. Select "DAI" as a payment option
4. Click through to pay
    1. Confirm two MetaMask transactions
5. Go to `/admin` to confirm the bookings; verify `invoice` call completes.

Which devices did you test on?
- Chrome on Mac
## REVIEWERS:
Check against these principles:

### High level
Does this code need to be written?
What are the alternatives?
Will this implementation become a support issue?
How much error margin does this solution have?

### Code
* Does the code follow industry standards?
JS: https://github.com/airbnb/javascript
React: https://github.com/airbnb/javascript/tree/master/react
https://github.com/vasanthk/react-bits
Documentation headers: http://usejsdoc.org/index.html

* Is there duplicated code? Can it be refactored into a shared method?
* Is the code consistent with our project?
* Are there unit tests? Do they test the states?
* Is the person refactoring another developer's code? If possible, did the original developer approve?

### Variables/Naming:
* Would the variable type led to future edge cases?
* Are the variable naming clear? Would the value contain something other than what the name describes.

### Security
* Can this be hacked or abused by the user?

## Further Work
1. More payment types
2. Backend initiation of the `collect` call

## Learnings
The [`using...for`](https://solidity.readthedocs.io/en/v0.5.2/contracts.html#using-for) keyword in Solidity allows contract-scoped decoration of arbitrary types; this is useful when you want contract interactions to be both [diligently-checked](https://github.com/woeltjen/unipay/blob/5c99fdebbe360efdc7a8bd887e6c9bdd1935c6f6/contracts/Unipay.sol#L179-L241) and yet [concise](https://github.com/woeltjen/unipay/blob/5c99fdebbe360efdc7a8bd887e6c9bdd1935c6f6/contracts/Unipay.sol#L283-L286)
